### ISSUE TITLE
[FEA] Support reading from cloud storage URLs

### ISSUE THREAD

**Issue by randerzander** [rapidsai/dask-cudf#109]
Dask can [read from cloud storage & HDFS](http://docs.dask.org/en/latest/remote-data-services.html) but reads into Pandas DataFrames. Today we can get that into dask_cudf with map_partitions and cudf.from_pandas, but that is slow and would be painful for users.

dask-cudf should support reading from S3, Google Cloud Storage, and Azure Data Lake Storage and passing directly into cudf file readers.

**Comment by mrocklin** [comment]
Yeah, we can do that, at least to the extent that cudf/cuio's readers can accept Python `bytes` objects as inputs.  I believe that this is true for `cudf.read_csv` at least today.

This might be as easy as adding a line like the following to dask-cudf

https://github.com/dask/dask/blob/6ecbb0cae94fd3c12e799b5c9721dd373117f2c7/dask/dataframe/io/csv.py#L495

```python
read_csv = make_reader(pd.read_csv, 'read_csv', 'CSV')
```

My guess is that we'll find minor differences between `pd.read_csv` and `cudf.read_csv` here and will need to resolve them, in principle everything should work ok here.

If that doesn't work then we could use [read_bytes](http://docs.dask.org/en/latest/bytes.html#dask.bytes.read_bytes) to give delayed objects for bytes values and then map `cudf.read_csv` over those bytes objects as appropriate ourselves.

cc @quasiben who has been doing read_csv work recently.  

@randerzander I'm curious how much of a priority you see this work relative to other things like discovering and fixing small usability bugs.

**Comment by randerzander** [comment]
This one is a must, given the push for YARN & Kubernetes deployments.

Note: I added HDFS above.

I'll defer commenting on priority vs "small usability bugs" until we've listed more of them.

**Comment by mrocklin** [comment]
We should be able to test this on local data.  I'll propose the following solution:

```python
def read_csv(path, **kwargs):
    if '://' in path:
        func = make_reader(cudf.read_csv, 'read_csv', 'CSV')
        return func(path, **kwargs)
    else:
        normal operation
```

Then I would test with local data and `'file:///path/to/mydata.*.csv'`.  If that works then I would expect s3/gcs/adl/hdfs to work just the same.

**Comment by mrocklin** [comment]
cc @beberg

**Comment by mrocklin** [comment]
This is in progress at https://github.com/rapidsai/dask-cudf/pull/122 , which is currently depending on https://github.com/rapidsai/cudf/pull/1066

**Comment by mrocklin** [comment]
Fixed in #122 by @quasiben 

```python
In [1]: import dask_cudf

In [2]: df = dask_cudf.read_csv('gcs://anaconda-public-data/nyc-taxi/csv/2015/yellow_*.csv')

In [3]: df
Out[3]: <dask_cudf.DataFrame | 273 tasks | 91 npartitions>

In [4]: df.head()
Out[4]: <cudf.DataFrame ncols=19 nrows=5 >

In [5]: _.to_pandas()
Out[5]:
   VendorID tpep_pickup_datetime  ... improvement_surcharge  total_amount
0         2  2015-01-15 19:05:39  ...                   0.3         17.05
1         1  2015-01-10 20:33:38  ...                   0.3         17.80
2         1  2015-01-10 20:33:38  ...                   0.3         10.80
3         1  2015-01-10 20:33:39  ...                   0.3          4.80
4         1  2015-01-10 20:33:39  ...                   0.3         16.30

[5 rows x 19 columns]

In [6]: df.dtypes
Out[6]:
VendorID                          int64
tpep_pickup_datetime     datetime64[ms]
tpep_dropoff_datetime    datetime64[ms]
passenger_count                   int64
trip_distance                   float64
pickup_longitude                float64
pickup_latitude                 float64
RateCodeID                        int64
store_and_fwd_flag                int32
dropoff_longitude               float64
dropoff_latitude                float64
payment_type                      int64
fare_amount                     float64
extra                           float64
mta_tax                         float64
tip_amount                      float64
tolls_amount                    float64
improvement_surcharge           float64
total_amount                    float64
dtype: object
```
